P2PEnabledConference.js

```diff
@@ -0,0 +1,726 @@
+                this._rejectIncomingCall(jingleSession);
+            }
+        } else {
+            // Let the JitsiConference deal with the JVB session
+            super.onIncomingCall(jingleSession, jingleOffer, now);
+        }
+    }
+
+    /**
+     * @inheritDoc
+     * @override
+     */
+    onMemberJoined (jid, nick, role, isHidden) {
+        super.onMemberJoined(jid, nick, role, isHidden);
+
+        this._startStopP2PSession();
+    }
+
+    /**
+     * @inheritDoc
+     * @override
+     */
+    onMemberLeft (jid) {
+        super.onMemberLeft(jid);
+
+        this._startStopP2PSession();
+    }
+
+    /**
+     * @inheritDoc
+     * @override
+     */
+    onRemoteTrackAdded (track) {
+        if (track.isP2P && !this.p2pEstablished) {
+            logger.info(
+                "Trying to add remote P2P track, when not in P2P - IGNORED");
+        } else if (!track.isP2P && this.p2pEstablished) {
+            logger.info(
+                "Trying to add remote JVB track, when in P2P - IGNORED");
+        } else {
+            super.onRemoteTrackAdded(track);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     * @override
+     */
+    onTransportInfo (jingleSession, transportInfo) {
+        if (this.peerToPeerSession === jingleSession) {
+            logger.info("Doing set transport-info");
+            this.peerToPeerSession.addIceCandidates(transportInfo);
+        }
+    }
+
+    /**
+     * Manually starts new P2P session (should be used only in the tests).
+     */
+    startPeer2PeerSession() {
+        const peers = this.getParticipants();
+        // Start peer to peer session
+        if (peers.length > 0) {
+            const peerJid = peers[0].getJid();
+            this._startPeer2PeerSession(peerJid);
+        } else {
+            logger.error("No participant to start the P2P session with !");
+        }
+    }
+
+    /**
+     * Manually stops the current P2P session (should be used only in the tests)
+     */
+    stopPeer2PeerSession() {
+        this._stopPeer2PeerSession();
+    }
+}
+
+/**
+ * This is a fake {@link ChatRoom} passed to the P2P {@link JingleSessionPC}
+ * in order to capture events emitted on it's event emitter (Jingle session uses
+ * chat room's emitter to send events).
+ */
+class FakeChatRoomLayer {
+
+    /**
+     * Creates new <tt>FakeChatRoomLayer</tt>
+     * @param p2pConference parent <tt>P2PEnabledConference</tt> instance
+     */
+    constructor(p2pConference) {
+
+        /**
+         * @type P2PEnabledConference
+         */
+        this.p2pConf = p2pConference;
+
+        /**
+         * See whatever docs are provided in
+         * the {@link ChatRoom#connectionTimes}.
+         * @type {Array}
+         */
+        this.connectionTimes = [];
+
+        /**
+         * Maps options of the original <tt>ChatRoom</tt>
+         */
+        this.options = p2pConference.room.options;
+        if (!this.options) {
+            logger.error("ChatRoom.options are undefined");
```

Comment:
It can happen when someone breaks the code. Just trying to detect this early and indicate what is missing. I could wait for some inner piece of code to break, but given that compiled/minimised code will provide "a, b, c" names it may be hard to guess anything out of it.